### PR TITLE
Support for inserting additional subsampling in conformer encoder

### PR DESCRIPTION
**Pull request by shan18**
# What does this PR do?

Added support for specifying an additional subsampling layer in conformer encoder. The user can specify the type and the position of the subsampling layer.

**Collection**: asr

# Changelog 
- New `SubsamplingReductionModule` class in `nemo/collections/asr/parts/submodules/subsampling.py`
- Updated the `ConformerEncoder` in `nemo/collections/asr/modules/conformer_encoder.py` to accept the new reduction parameters for adding a new subsampling layer in the desired position.
- Updated the _transducer-related configs_ with reduction parameters.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Everyone in the NeMo ASR team.

### CODE REVIEW

File: docs/source/asr/speech_intent_slot/configs.rst [hunk 1]
```diff
@@ -88,6 +88,11 @@ An example config for the model can be:
       subsampling_factor: 4 # must be power of 2
       subsampling_conv_channels: -1 # -1 sets it to d_model
 
+      # Reduction params
```
Comment (VahidooX) [inline docs/source/asr/speech_intent_slot/configs.rst:91]: Maybe adding a comment what is the purpose of reduction?

File: docs/source/asr/speech_intent_slot/configs.rst [hunk 2]
```diff
@@ -88,6 +88,11 @@ An example config for the model can be:
...
+      reduction: null # pooling, subsampling, or null
+      reduction_position: -2 # Encoder block index or -1 for subsampling at the end of encoder
```
Comment (VahidooX) [inline docs/source/asr/speech_intent_slot/configs.rst:93]: What does -2 does?
How about using null to disable it?

File: docs/source/asr/speech_intent_slot/configs.rst [hunk 3]
```diff
@@ -88,6 +88,11 @@ An example config for the model can be:
       subsampling_factor: 4 # must be power of 2
       subsampling_conv_channels: -1 # -1 sets it to d_model
 
+      # Reduction params
+      reduction: null # pooling, subsampling, or null
```
Comment (VahidooX) [inline docs/source/asr/speech_intent_slot/configs.rst:92]: What would be the subsampling approach?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 4]
```diff
@@ -60,6 +64,12 @@ class ConformerEncoder(NeuralModule, StreamingEncoder, Exportable):
             Defaults to 4.
         subsampling_conv_channels (int): the size of the convolutions in the subsampling module
             Defaults to -1 which would set it to d_model.
+        reduction (str, Optional): the method of reduction, choices=['pooling', 'subsampling']. If no value
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:67]: What do you mean by not passing any value? do you mean None?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 5]
```diff
@@ -252,6 +266,16 @@ def __init__(
         else:
             self.pre_encode = nn.Linear(feat_in, d_model)
 
+        # Reduction
+        if reduction and reduction_factor > 1:
+            self.reduction_subsampling = SubsamplingReductionModule(
+                reduction=reduction, d_model=d_model, reduction_factor=reduction_factor,
+            )
+            self.reduction_position = reduction_position
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:274]: Let's check that reduction_position has a valid value and raise an error otherwise.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 6]
```diff
@@ -384,6 +408,46 @@ def forward(self, audio_signal, length, cache_last_channel=None, cache_last_time
...
+            audio_signal, length = self.reduction_subsampling(x=audio_signal, lengths=length)
+
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+
+        if cache_last_channel is not None:
+            return audio_signal, length, cache_last_channel_next, cache_last_time_next
+        else:
+            return audio_signal, length
+
+    def _create_masks(self, audio_signal, max_audio_length, length, cache_last_channel=None):
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:450]: First part of the code related to caching should not be inside create_masks. It should just creates the maskings.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 7]
```diff
@@ -384,6 +408,46 @@ def forward(self, audio_signal, length, cache_last_channel=None, cache_last_time
         max_audio_length = audio_signal.size(1)
 
         # Create the self-attention and padding masks
+        audio_signal, max_audio_length, pad_mask, att_mask, pos_emb, cache_last_channel_next = self._create_masks(
```
Comment (VahidooX) [inline nemo/collections/asr/modules/conformer_encoder.py:411]: What is the benefit of creating this method of _create_masks? Is it going to be used anywhere else?

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 2 above)
Comment (bmwshop) [reply docs/source/asr/speech_intent_slot/configs.rst:93]: Agree with @VahidooX ; it's not clear what -2 means.

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 3 above)
Comment (bmwshop) [reply docs/source/asr/speech_intent_slot/configs.rst:92]: +1 on explaining the approaches. Suggest using None instead of null

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 2 above)
Comment (shan18) [reply docs/source/asr/speech_intent_slot/configs.rst:93]: - -2 means no reduction is applied.
- -1 means that reduction is applied at the end of encoder (after the output projection layer)
- Integers 0-17 specify the block index after which the subsampling should be applied.

Having null instead for this value also works. I'll change it.

As for checking the bounds, should I add an assert statement to ensure the value is either None or remains between [-1, len(self.layers)]?

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 3 above)
Comment (shan18) [reply docs/source/asr/speech_intent_slot/configs.rst:92]: The subsampling approach is the initial 4x subsampling that we apply to the encoder. It was there already.
What I added are the reduction parameters which add another 2x subsampling layer in the specified position.

Regarding the null values, across all the configs in NeMo, I have seen the values null specified in yaml files. When we use these files in python, the null values automatically get converted to None type.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 4 above)
Comment (shan18) [reply nemo/collections/asr/modules/conformer_encoder.py:67]: If the reduction parameter is not specified in the config then no such parameter is passed while creating an object of this class so the values are by default None.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 5 above)
Comment (shan18) [reply nemo/collections/asr/modules/conformer_encoder.py:274]: does an assert statement to ensure that the value remains between `[-1, len(self.layers)]` sound good?

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 7 above)
Comment (shan18) [reply nemo/collections/asr/modules/conformer_encoder.py:411]: Yes, if we apply subsampling in between the encoder blocks then we need to calculate the masks again because the length of the audio signal will be reduced. So for this purpose, I put it in a separate function.

See line 434. I call this function there again.

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 2 above)
Comment (VahidooX) [reply docs/source/asr/speech_intent_slot/configs.rst:93]: Yes, that is the bound to get checked.

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 3 above)
Comment (VahidooX) [reply docs/source/asr/speech_intent_slot/configs.rst:92]: Looks like the reduction just supports striding. So I suggest to use from ["pooling", "striding"] and raise a "not supported" error if something else is passed.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 5 above)
Comment (VahidooX) [reply nemo/collections/asr/modules/conformer_encoder.py:274]: Yes, sounds good.

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 1 above)
Comment (shan18) [reply docs/source/asr/speech_intent_slot/configs.rst:91]: Added

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 2 above)
Comment (shan18) [reply docs/source/asr/speech_intent_slot/configs.rst]: Added a bound check in the new commit

File: docs/source/asr/speech_intent_slot/configs.rst (hunk 3 above)
Comment (shan18) [reply docs/source/asr/speech_intent_slot/configs.rst]: Have added that check in `subsampling.py`, inside the `SubsamplingReductionModule` class

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 6 above)
Comment (shan18) [reply nemo/collections/asr/modules/conformer_encoder.py]: Pushed the code with caching outside create_masks

**Review by VahidooX — APPROVED** [review]

**Review by VahidooX — APPROVED** [review]

**Review by bmwshop — APPROVED** [review]
